File: Deep_Learning/CNN/cnn.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Feb  6 09:46:26 2025
@author: swarnabhaghosh
"""

# Convolutional Neural Network (CNN)
import os
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense
from tensorflow.keras.preprocessing.image import ImageDataGenerator

# Automatically detect script directory (portable)
script_dir = os.path.dirname(os.path.abspath(__file__))
os.chdir(script_dir)

# Initialize CNN
classifier = Sequential()

# Convolution
classifier.add(Conv2D(filters=32, kernel_size=(3, 3), input_shape=(64, 64, 3), activation='relu'))

# Pooling
classifier.add(MaxPooling2D(pool_size=(2, 2)))

# Adding 2nd convolutional layer
classifier.add(Conv2D(filters=32, kernel_size=(3, 3), activation='relu'))
classifier.add(MaxPooling2D(pool_size=(2, 2)))

# Flattening
classifier.add(Flatten())

# Fully Connected Layers
classifier.add(Dense(units=128, activation='relu'))
classifier.add(Dense(units=1, activation='sigmoid'))  # Sigmoid for binary classification

# Compile the CNN
classifier.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Data Augmentation
train_datagen = ImageDataGenerator(rescale=1./255, shear_range=0.2, zoom_range=0.2, horizontal_flip=True)
test_datagen = ImageDataGenerator(rescale=1./255)

# Load Data
training_set = train_datagen.flow_from_directory(os.path.join(script_dir, 'dataset/training_set'), target_size=(64, 64), batch_size=32, class_mode='binary')
test_set = test_datagen.flow_from_directory(os.path.join(script_dir, 'dataset/test_set'), target_size=(64, 64), batch_size=32, class_mode='binary')

# Train the CNN
classifier.fit(training_set, steps_per_epoch=len(training_set), epochs=25, validation_data=test_set, validation_steps=len(test_set))

# Save the model
classifier.save(os.path.join(script_dir, 'cnn_model.h5'))  # Save model in the same directory
print("✅ CNN Model Training Completed!")

File: Deep_Learning/CNN/cnn.R
# Convolutional Neural Network (CNN) in R using Keras & TensorFlow

# Install and Load Required Packages
if (!require(keras)) install.packages("keras")
if (!require(tensorflow)) install.packages("tensorflow")

library(keras)
library(tensorflow)

# Install TensorFlow (if not already installed)
install_tensorflow()

# Data Preprocessing
train_datagen <- image_data_generator(
  rescale = 1/255,
  shear_range = 0.2,
  zoom_range = 0.2,
  horizontal_flip = TRUE
)

test_datagen <- image_data_generator(rescale = 1/255)

training_set <- flow_images_from_directory(
  "dataset/training_set",
  train_datagen,
  target_size = c(64, 64),
  batch_size = 32,
  class_mode = "binary"
)

test_set <- flow_images_from_directory(
  "dataset/test_set",
  test_datagen,
  target_size = c(64, 64),
  batch_size = 32,
  class_mode = "binary"
)

# Building the CNN Model
cnn_model <- keras_model_sequential() %>%
  
  # Convolution Layer 1
  layer_conv_2d(filters = 32, kernel_size = c(3, 3), activation = "relu", input_shape = c(64, 64, 3)) %>%
  layer_max_pooling_2d(pool_size = c(2, 2)) %>%
  
  # Convolution Layer 2
  layer_conv_2d(filters = 32, kernel_size = c(3, 3), activation = "relu") %>%
  layer_max_pooling_2d(pool_size = c(2, 2)) %>%
  
  # Flattening
  layer_flatten() %>%
  
  # Fully Connected Layer
  layer_dense(units = 128, activation = "relu") %>%
  layer_dropout(rate = 0.5) %>%
  
  # Output Layer (Binary Classification)
  layer_dense(units = 1, activation = "sigmoid")

# Compile the CNN Model
cnn_model %>% compile(
  optimizer = "adam",
  loss = "binary_crossentropy",
  metrics = c("accuracy")
)

# Train the CNN Model
history <- cnn_model %>% fit(
  training_set,
  steps_per_epoch = 250,
  epochs = 25,
  validation_data = test_set,
  validation_steps = 63
)

# Save the Model
save_model_hdf5(cnn_model, "cnn_model.h5")

# Print Summary
summary(cnn_model)
